src/Storages/MergeTree/MergeTreeIndexHypothesisMergedCondition.cpp

```diff
@@ -16,199 +21,371 @@ namespace ErrorCodes
+    if (func)
+        return &func->name;
 
-    for (const auto & group : cnf.getStatements())
-    {
-        if (group.size() == 1)
-        {
-            hypotheses_data.push_back(group);
-            CNFQuery::AtomicFormula atomic_formula = *group.begin();
-            CNFQuery::AtomicFormula atom{atomic_formula.negative, atomic_formula.ast->clone()};
-            pushNotIn(atom);
-            assert(!atom.negative);
-
-            const auto * func = atom.ast->as<ASTFunction>();
-            if (func && relations.contains(func->name))
-                compare_hypotheses_data.push_back(atom.ast);
-        }
-    }
+    return nullptr;
+}
+
+const std::string * functionName(const Analyzer::CNF::AtomicFormula & atom)
+{
+    const auto * function = atom.node_with_hash.node->as<FunctionNode>();
+
+    if (function)
+        return &function->getFunctionName();
 
-    index_to_compare_atomic_hypotheses.push_back(compare_hypotheses_data);
-    index_to_atomic_hypotheses.push_back(hypotheses_data);
+    return nullptr;
 }
 
-void MergeTreeIndexhypothesisMergedCondition::addConstraints(const ConstraintsDescription & constraints_description)
+bool isFunction(const CNFQuery::AtomicFormula & atom)
 {
-    auto atomic_constraints_data = constraints_description.getAtomicConstraintData();
-    for (const auto & atomic_formula : atomic_constraints_data)
-    {
-        CNFQuery::AtomicFormula atom{atomic_formula.negative, atomic_formula.ast->clone()};
-        pushNotIn(atom);
-        atomic_constraints.push_back(atom.ast);
-    }
+    return atom.ast->as<ASTFunction>() != nullptr;
 }
 
-/// Replaces < -> <=, > -> >= and assumes that all hypotheses are true then checks if path exists
-bool MergeTreeIndexhypothesisMergedCondition::alwaysUnknownOrTrue() const
+bool isFunction(const Analyzer::CNF::AtomicFormula & atom)
+{
+    return atom.node_with_hash.node->as<FunctionNode>() != nullptr;
+}
+
+const auto & functionArguments(const CNFQuery::AtomicFormula & atom)
+{
+    return atom.ast->as<ASTFunction &>().arguments->children;
+}
+
+const auto & functionArguments(const Analyzer::CNF::AtomicFormula & atom)
 {
-    ASTs active_atomic_formulas(atomic_constraints);
-    for (const auto & hypothesis : index_to_compare_atomic_hypotheses)
+    return atom.node_with_hash.node->as<FunctionNode &>().getArguments().getNodes();
+}
+
+template <typename Node>
+class IndexImpl : public MergeTreeIndexhypothesisMergedCondition::IIndexImpl
+{
+public:
+    IndexImpl(const SelectQueryInfo & query_info, const ConstraintsDescription & constraints, ContextPtr context_)
+        : context(std::move(context_))
+        , table_expression(query_info.table_expression)
     {
-        active_atomic_formulas.insert(
-            std::end(active_atomic_formulas),
-            std::begin(hypothesis),
-            std::end(hypothesis));
+        auto ast_filter_node = buildFilterNode(query_info.query);
```

Comment:
I suspect that this can be wrong in very edge cases because this filter AST part in Analyzer is not valid.

Reply:
Can you explain a bit more? Why wouldn't be valid?

Reply:
As far as I understand we use a query tree that can be different from AST in `query_info.query` because of optimizations. So in the case of `allow_experimental_analyzer` you want to use `query_info.query_tree` to build a filter.

Reply:
@antonio2368 I spent some time and understood that the problem is more complex. The filters in the query tree are also not those that we use. For example, it misses row policy filters which are added in the Planner. In the case of a new analyzer, the correct filter is stored in the `SelectQueryInfo::filter_actions_dag`. So you need to use it.